File: uni_dist/reduce_scatter.py
# write an all gather function that can use either nccl or mpi.
#  
# all_gather.py

import torch
import torch.distributed as dist
from mpi4py import MPI
from typing import List, Optional, Union
from .request import Request
from .process_groups import ProcessGroups
import numpy as np
from .utils import _torch_to_mpi

def ring_reduce_scatter_mpi(output_tensor: torch.Tensor,
                        input_tensor: torch.Tensor,
                        group: Optional[MPI.Comm] = None,
                        async_op: bool = False, 
                        op=torch.add):
    """
    Performs a ring-based reduce-scatter using torch tensors.
    
    Each process holds a 2D tensor 'sendbuf' of shape (P, N), where P is the number 
    of processes and N is the block size. Each row of the tensor corresponds to a block.
    The goal is to reduce the i-th block (row) from all processes (using the provided op)
    so that at the end, each process gets the fully reduced block corresponding to its 
    assigned segment.
    
    The algorithm performs P-1 steps. At each step, each process:
      - Sends a designated block to its right neighbor.
      - Receives a block from its left neighbor.
      - Immediately reduces the received block into its local copy.
      
    After the loop, the fully reduced block is at index (rank - (P-1)) % P.
    """
    assert not async_op, "non-blocking primitives not supported"
    comm = MPI.COMM_WORLD if group is None else group
    rank = comm.Get_rank()
    size = comm.Get_size()
    
    assert input_tensor.dim() == 1, "input tensor should be 1D"
    assert output_tensor.dim() == 1, "output tensor should be 1D"

    sendbuf = input_tensor.view(size, -1)

    # Ensure that there is one block per process.
    assert sendbuf.size(0) == size, "sendbuf must have one block per process"
    block_size = sendbuf.size(1)
    
    # Work on a local copy of the buffer.
    buf = sendbuf.clone()
    tmp = torch.empty(block_size, dtype=sendbuf.dtype, device=sendbuf.device)
    
    # Perform size-1 steps of the ring reduce-scatter.
    for step in range(size - 1):
        # Determine the block indices.
        send_idx = (rank - step) % size
        recv_idx = (rank - step - 1) % size
        
        # Identify neighbors in the ring.
        dest = (rank + 1) % size
        source = (rank - 1 + size) % size
        
        # Copy the block to be sent.
        send_data = buf[send_idx] #.clone()
        
        # Send the designated block to the right and receive a block from the left.
        torch.cuda.current_stream().synchronize()
        comm.Sendrecv(sendbuf=send_data, dest=dest, sendtag=0,
                      recvbuf=tmp, source=source, recvtag=0)
        
        # Reduce the received block into our local block.
        buf[recv_idx] = op(buf[recv_idx], tmp)
    
    # The final, fully reduced block is at this index.
    final_idx = (rank - (size - 1)) % size
    output_tensor.copy_(buf[final_idx])


def _reduce_scatter(
    output_tensor: torch.Tensor,
    input_tensor: torch.Tensor,
    group: Optional[Union[dist.ProcessGroup, MPI.Comm]] = None,
    async_op: bool = False,
    directly_call_mpi: bool = False
) -> Optional[Request]:

    # Case 1: torch.distributed.ProcessGroup
    if group is None or isinstance(group, dist.ProcessGroup):
        # Delegate to torch.distributed.all_gather
        request = dist.reduce_scatter_tensor(output_tensor, 
                                             input_tensor, 
                                             group=group, 
                                             async_op=async_op)
    # Case 2: mpi4py.MPI.Comm
    elif isinstance(group, MPI.Comm):
        # make sure that the cpu is synchronized with the current stream
        if not directly_call_mpi:
            request = ring_reduce_scatter_mpi(output_tensor, input_tensor, group, async_op)
        else:
            request = group.Reduce_scatter_block(input_tensor, output_tensor)
    else:
        raise TypeError(
            f"Unsupported group type: {type(group)}. "
            "Expected torch.distributed.ProcessGroup or mpi4py.MPI.Comm."
        )
    return request 

def reduce_scatter_2D(output_tensor: torch.Tensor,
    input_tensor: torch.Tensor,
    group: Optional[ProcessGroups] = None,
    async_op: bool = False):

    assert not async_op, "Non blocking version not implemented"

    assert input_tensor.dim() == 1 and output_tensor.dim() == 1, "all_gather_2D only admits 1D tensors"
    intra_node_group_size, inter_node_group_size = group.get_world_size()

    # step 1: on-device permutation 
    world_size = inter_node_group_size * intra_node_group_size
    output_msg_size = input_tensor.size(0) // world_size
    input_splits = torch.split(input_tensor, split_size_or_sections=output_msg_size)
    permuted_tensors = []
    for i in range(intra_node_group_size):
        idxes = list(np.arange(i, world_size, intra_node_group_size))
        permuted_tensors.extend([input_splits[idx] for idx in idxes])
    input_permuted = torch.cat(permuted_tensors)

    # Step-2 intra-node reduce-scatter
    output_intermediate = torch.empty(input_permuted.size(0) // intra_node_group_size, 
                                      device=input_tensor.device, 
                                      dtype=input_tensor.dtype)
    _reduce_scatter(output_intermediate, input_permuted, group.get_inner_group(), async_op=False)

    # Step-2 inter-node 
    _reduce_scatter(output_tensor, output_intermediate, group.get_outer_group(), async_op=False)

File: uni_dist/src/reduce_scatter.cpp
#include "reduce_scatter.h"
#include "utils.h"
#include <torch/extension.h>
#include <torch/torch.h>
#include <hip/hip_runtime.h>
#include <cassert>
#include <cmath>
#include <ATen/hip/HIPContext.h>

// Performs a recursive-halving reduce-scatter on GPU tensors.
//  - output: HIP device pointer where the reduced block (of block_size elements) will be stored.
//  - input: HIP device pointer to a contiguous buffer of (P * block_size) elements.
//  - total_elems: total number of elements (must equal P * block_size).
//  - comm: MPI communicator (default MPI_COMM_WORLD).
// The reduction operation is elementwise addition.
void recursiveHalvingReduceScatterGPU(float* output, 
    const float* input, 
    int total_elems,
    float* buf, // same as size of input 
    float* recv_buf, // same as size of input 
    MPI_Comm comm) {
    
    int rank, size;
    MPI_Comm_rank(comm, &rank);
    MPI_Comm_size(comm, &size);

    assert(total_elems % size == 0 && "Input tensor size must be divisible by number of processes");
    int block_size = total_elems / size;
    auto hip_stream = at::hip::getCurrentHIPStream();

    // copy the input into buf
    HIP_CHECK(hipMemcpyAsync(buf, 
        input, 
        total_elems * sizeof(float), 
        hipMemcpyDeviceToDevice,
        hip_stream));

    
    int max_count = (size / 2) * block_size;
    
    // Number of rounds = log2(size); assumes size is a power of 2.
    int rounds = static_cast<int>(std::log2(size));
    int current_blocks = size;
    // curr_ptr points to the beginning of the "active" region within buf.
    float* curr_ptr = buf;

    hipEvent_t stream_sync_event;
    HIP_CHECK(hipEventCreateWithFlags(&stream_sync_event, hipEventDisableTiming));

    for (int r = 0; r < rounds; ++r) {
        // In round r, the group size is:
        int group_size = size / (1 << r);
        // The current buffer holds 'current_blocks' contiguous blocks.
        int half = current_blocks / 2;
        // Number of elements to send/receive in this round.
        int count = half * block_size;

        if ((rank % group_size) < (group_size / 2)) {
            // Lower half: keep the lower half and send the upper half.
            int partner = (rank + half) % size;
            float* send_ptr = curr_ptr + half * block_size;
            // Record an event on the hip stream.
            HIP_CHECK(hipEventRecord(stream_sync_event, hip_stream));
            // Wait for the copy to complete.
            HIP_CHECK(hipEventSynchronize(stream_sync_event));

            MPI_Sendrecv(send_ptr, count, MPI_FLOAT, partner, r,
                         recv_buf, count, MPI_FLOAT, partner, r,
                         comm, MPI_STATUS_IGNORE);
            // Reduce the received data into the lower half.
            vectorAdd(curr_ptr, recv_buf, count, hip_stream);
            current_blocks = half;
        } else {
            // Upper half: keep the upper half and send the lower half.
            int partner = (rank - half + size) % size;
            float* send_ptr = curr_ptr;  // lower half to be sent.
            // Record an event on the hip stream.
            HIP_CHECK(hipEventRecord(stream_sync_event, hip_stream));
            // Wait for the copy to complete.
            HIP_CHECK(hipEventSynchronize(stream_sync_event));
            MPI_Sendrecv(send_ptr, count, MPI_FLOAT, partner, r,
                         recv_buf, count, MPI_FLOAT, partner, r,
                         comm, MPI_STATUS_IGNORE);
            float* kept_ptr = curr_ptr + half * block_size;
            vectorAdd(kept_ptr, recv_buf, count, hip_stream);
            curr_ptr = kept_ptr;
            current_blocks = half;
        }
    }

    // After log2(size) rounds, curr_ptr points to a single block (of block_size elements).
    HIP_CHECK(hipMemcpyAsync(output, 
        curr_ptr, 
        block_size * sizeof(float), 
        hipMemcpyDeviceToDevice, 
        hip_stream));

    HIP_CHECK(hipEventDestroy(stream_sync_event));
}

void ringReduceScatterGPU(float* output, 
    const float* input, 
    int total_elems, 
    float* d_buf, // same as size of input
    float* d_send, // same as size of output
    float* d_tmp, // same as size of output
    MPI_Comm comm
) {
    int rank, size;
    MPI_Comm_rank(comm, &rank);
    MPI_Comm_size(comm, &size);
    
    assert(total_elems % size == 0 && "Input tensor size must be divisible by number of processes");
    int block_size = total_elems / size;

    auto hip_stream = at::hip::getCurrentHIPStream();
    hipEvent_t stream_sync_event;
    
    HIP_CHECK(hipMemcpyAsync(d_buf, 
        input, 
        total_elems * sizeof(float), 
        hipMemcpyDeviceToDevice,
        hip_stream));

    HIP_CHECK(hipEventCreateWithFlags(&stream_sync_event, hipEventDisableTiming));

    for (int step = 0; step < size - 1; step++) {
        // Compute block indices.
        int send_idx = (rank - step - 1 + size) % size;
        int recv_idx = (rank - step - 2 + size) % size;
        int dest     = (rank + 1) % size;
        int source   = (rank - 1 + size) % size;

        // Record an event on the hip stream.
        HIP_CHECK(hipEventRecord(stream_sync_event, hip_stream));
        // Wait for the copy to complete.
        HIP_CHECK(hipEventSynchronize(stream_sync_event));

        // Send the block to the right neighbor and receive from the left neighbor.
        MPI_Sendrecv(d_buf + send_idx * block_size, block_size, MPI_FLOAT, dest, 0,
                    d_tmp,  block_size, MPI_FLOAT, source, 0,
                    comm, MPI_STATUS_IGNORE);

        // Reduce the received block into the block at recv_idx.
        vectorAdd(d_buf + recv_idx * block_size, d_tmp, block_size, hip_stream);
    }
     
    HIP_CHECK(hipMemcpyAsync(output, 
        d_buf + rank * block_size, 
        block_size * sizeof(float), 
        hipMemcpyDeviceToDevice,
        hip_stream));

    HIP_CHECK(hipEventDestroy(stream_sync_event));
}

